text/0113-keyspace-level-gc.md

```diff
@@ -0,0 +1,83 @@
+the GC safe point value of the data key and execute the GC logic.
+
+
+## Concepts of GC management type:
+
+1. Global GC:
+    - Represents the previous default GC logic; there is a TiDB calculate the global GC safe point for the whole cluster.
+    - The default GC management type for keyspace is Global GC,
+2. Keyspace level GC:
+    - Indicates that the keyspace will advance its own GC safe point.
+    - It is possible and only possible to set gc_management_type = keyspace_level_gc when PD creates keyspaces.
+    - The keyspace which already set gc_management_type = keyspace_level_gc, 'gc_management_type' it can not be updated 
+      to "global_gc".
+    - Keyspace GC-related data: min start ts, GC safe point,service safe point of keyspace have their own etcd path 
+      with keyspace prefix in PD. 
+    - Therefore, the keysapce can calculate the GC by itself.
+
+
+## Implementation in TiKV
+
+1. Get keyspace meta and keyspace level GC safe point:
+    - New KeyspaceMetaWatchService : Watch keyspace meta etcd path to get the keyspace meta and put it 
+      in cache keyspace_id_meta_map<u32, keyspacepb::KeyspaceMeta>.
+    - New KeyspaceLevelGCWatchService : Watch the etcd path of the keyspace GC safe point to get the GC safe point 
+      with keyspace level GC enabled, put it in cache keyspace_level_gc_map<u32, u64>.
+
+2. How to get GC safe point by mvcc key in Compaction Filter:
+![img.png](../media/keyspace-level-gc-get-gc-sp.png)
+
+3. How to determine if a keyspace uses a global GC safe point:
+![img.png](../media/keyspace-level-gc-is-global-gc.png)
+
+4. Use GC safe point to optimize trigger timing and assert judgment:
+   1. In the global GC, it will skip GC when GC safe point is 0 in create_compaction_filter. 
+      After supporting keyspace level GC, GC is skipped if global GC safe point is 0 or if no keyspace level GC is initialized.
+   2. In the global GC, check_need_gc function return false in create_compaction_filter.
+      After supporting keyspace level GC, if props.min_ts > global GC safe point and props.min_ts > all keyspace level GC safe point will return false.
+   3. In the global GC, assert( safe_point > 0 ) when new compaction filter.
+      After supporting keyspace level GC, the assert condition is whether global GC safe point > 0 or any keyspace level GC safe point has been initialized.
+
+5. Other non-GC logic that uses GC safe point does not currently have to support keyspace level GC.
+   1. check region consistency command: 
+      1. It needs check GC safe point on followers to ensure that the data to be checked on the follower is not GC. 
+      2. It doesn't support which keyspace with keyspace level GC enabled yet, if user request check consistency 
+      for the keyspace range, the user will get a "not supported" message.
+   2. GC safe point used in raftstore to trigger compaction when no valid split key can be found. 
```

Comment:
We can just write the design here. It doesn't matter that it has been implemented or not.

Reply:
I think it can be kept for now, Wait for other reviewers to review.

Reply:
Move it to PR describe https://github.com/tikv/tikv/pull/16808